colossalai/inference/core/request_handler.py

```diff
@@ -1,48 +1,210 @@
+    def _find_sequence(self, request_id: str) -> Sequence:
         """
-        Find the request by seq_id.
+        Find the request by request_id.
         """
+        for priority, lst in enumerate(self.waiting_list):
+            for seq in lst:
+                if seq.request_id == request_id:
+                    return seq, priority
+
+        if self.running_list.find_seq(request_id):
+            return seq, None
+
+        return None
+
+    def _sample(self, probs: torch.Tensor, logprobs: torch.Tensor, generation_config):
+        if generation_config.num_beams == 1:
+            if generation_config.do_sample:
+                sample_tokens = greedy_sample(generation_config, logprobs)
+            else:
+                sample_tokens = multinomial_sample(generation_config, probs)
+        else:
+            sample_tokens = beam_search_sample(generation_config, logprobs, is_prompt=not self.prefill_batch.is_empty)
+
+        return sample_tokens
+
+    def mark_finished(self, sequence: Sequence, generation_config):
+        if (
+            sequence.output_token_id[-1] == generation_config.eos_id
+            or sequence.output_len >= generation_config.max_output_len
+        ):
+            sequence.mark_finished()
 
     def check_unfinished_seqs(self) -> bool:
-        return self.waiting_list or self.running_list
+        return self._has_waiting() or not self.running_list.is_empty()
+
+    def search_tokens(self, generation_config, logits):
+        """
+        Sample tokens for finished requests.
+        """
+        # do logit processor
+        # NOTE: need to decide the granularity to process logits (sequence or batch)
+        for type in ["top_p", "top_k", "min_p"]:
+            if type in generation_config:
+                logits = logit_processor(type, logits)
+
+        # calculate probs
+        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
+        logprobs = torch.log_softmax(logits, dim=-1, dtype=torch.float)
+
+        # sample the next tokens
+        sample_tokens = self._sample(probs, logprobs, generation_config)
+
+        for idx, sample in enumerate(sample_tokens):
+            sequence = self.running_batch.sequences_set[idx]
+            sequence.output_token_id.append(sample)
+            self.mark_finished(sequence, generation_config)
+
+    def update(self):
+        """
+        Update current running list and done list
+        """
+        if not self.prefill_batch.is_empty:
+            self.running_list.decoding.extend(self.running_list.prefill)
+            self.running_batch.add_seqs(self.running_list.prefill)
+            self.running_list.prefill.clear()
+            self.running_batch.clear_batch()
+
+        for seq in self.running_batch.sequences_set:
+            if seq.check_finish():
+                self.done_list.append(seq)
+                self.running_list.remove(seq)
+                self.running_batch.sequences_set.remove(seq)
```

Comment:
You should call discard(seq) instead of remove(seq), and I think you can add a function in BatchInfo to release without using to sequences sets directly.

Reply:
fixed